### PR TITLE
tests: ensure CI test run generates proper junit output files for Jenkins reporting

### PR DESCRIPTION
**Pull request by trentm**
Fixes: #2443
Obsoletes: #2471

### Checklist

- [x] Implement code
- [x] do a full test run (with Edge, TAV, etc.) to ensure that still works: https://apm-ci.elastic.co/job/apm-agent-nodejs/job/apm-agent-nodejs-mbp/job/PR-2478/12/
- [x] Fix the output to dump the created TAP files so we have that TAP content for grepping in the log output when things fail.

### CODE REVIEW

**Comment by trentm** [comment]
Worked with tap-junit (https://apm-ci.elastic.co/job/apm-agent-nodejs/job/apm-agent-nodejs-mbp/job/PR-2478/5/), and I got 233539 test cases (as @v1v did at https://github.com/elastic/apm-agent-nodejs/pull/2471#issuecomment-982931952) so that seems to confirm this is a difference in tap-xunit vs tap-junit parsing of the TAP output.

**Comment by trentm** [comment]
The full test run attempt failed on a spurious error: https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-nodejs%2Fapm-agent-nodejs-mbp/detail/PR-2478/6/pipeline/169

```
Archive JUnit-formatted test results<1s
[2021-11-30T20:35:06.000Z] Recording test results
[2021-11-30T20:35:06.308Z] Remote call on apm-ci-immutable-ubuntu-1804-1638303653075530776 failed
Remote call on apm-ci-immutable-ubuntu-1804-1638303653075530776 failed
```

I'll try that again later.

**Comment by trentm** [comment]
> dump the TAP file content to stdout to have it in the log

That worked. For example look at logs for steps at https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-nodejs%2Fapm-agent-nodejs-mbp/detail/PR-2478/7/pipeline/1143
(Unrelated: I killed that build because it got stuck in "Report build result".)

**Comment by trentm** [comment]
I think there may now be a new issue. The "Reporting build status" step of these jobs is now taking **30+ minutes** as can be seen here: https://apm-ci.elastic.co/job/apm-agent-nodejs/job/apm-agent-nodejs-mbp/job/PR-2478/

Looking at https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-nodejs%2Fapm-agent-nodejs-mbp/detail/PR-2478/8/pipeline I believe the delay is here in `generate-build-data.sh`

```
[2021-11-30T23:48:37.491Z] + chmod 755 generate-build-data.sh
[2021-11-30T23:48:37.491Z] + ./generate-build-data.sh https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/apm-agent-nodejs-mbp/PR-2478/ https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/apm-agent-nodejs-mbp/PR-2478/runs/8 SUCCESS 6777552
...
[2021-11-30T23:48:49.850Z] INFO: fetchAndPrepareTestsInfo (see tests-info.json)
[2021-11-30T23:48:49.850Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/apm-agent-nodejs-mbp/PR-2478/runs/8/tests/?limit=10000000 -o tests-info.json
```

The `curl ...` at the end of this incomplete log is hanging or at least going slowly and taking much too long. When I try to do that same curl on my dev machine:

```
% curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/apm-agent-nodejs-mbp/PR-2478/runs/8/tests/?limit=10000000 -o tests-info.json
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 8640k    0 8640k    0     0  17962      0 --:--:--  0:08:12 --:--:--  6116
```

So far 9MB of JSON has been downloaded in ~8 minutes:

```
% du -sh tests-info.json
9.0M	tests-info.json
```

Is this because there are now 200k tests? Or is it because something in Jenkins is overloaded currently? @v1v: Ideas?

**Comment by trentm** [comment]
> The "Reporting build status" step of these jobs is now taking **30+ minutes**

And fails after ~30 minutes. So this is now a blocker.

I wonder if it is not feasible to the "/blue/rest/organizations/jenkins/pipelines/.../runs/.../tests/" Jenkins endpoint for such a larger number of tests. If this data isn't required, then perhaps we could skip having "tests-info.json". If it is required, then perhaps using the raw Junit files and ensuring those are added as artifacts to builds?

**Comment by v1v** [comment]
> Is this because there are now 200k tests? Or is it because something in Jenkins is overloaded currently? @v1v: Ideas?

That's normally the reason, https://github.com/elastic/apm-pipeline-library/pull/1417 is the PR I raised when I worked with the tap-to-junit-ext step as I discovered that such amount of tests caused some issues when fetching them.

In any case, this should not be a blocker since the last stage in any pipeline for reporting has been implemented in such a way that will not report a build failure:

image

**Comment by trentm** [comment]
> In any case, this should not be a blocker since the last stage in any pipeline for reporting has been implemented in such a way that will not report a build failure:

I disagree that this is not a blocker. It adds 30 minutes to the build times.

My attempt to download that Jenkins endpoint eventually failed after almost **2 hours**:

```
% curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/apm-agent-nodejs-mbp/PR-2478/runs/8/tests/?limit=10000000 -o tests-info.json
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 54.3M    0 54.3M    0     0   8225      0 --:--:--  1:55:31 --:--:--     0
curl: (56) LibreSSL SSL_read: SSL_ERROR_SYSCALL, errno 60
```

by that point it had downloaded 54 MB:

```
% du -sh tests-info.json
 54M	tests-info.json
```

I think it failed when I closed my laptop at the end of my workday.
(I also wonder at the possible CPU load being put on the Jenkins infra for calls like this.)

**Comment by v1v** [comment]
> I disagree that this is not a blocker. It adds 30 minutes to the build times.

I disagree then :) a blocker implies no way to do things, while adding 30 minutes might not be ideal but still things keep working. 

As a workaround, If you'd like to speed up your development, you could comment out the `notifyBuildResult` step in the Jenkinsfile. 


> (I also wonder at the possible CPU load being put on the Jenkins infra for calls like this.)

No other projects run 200.000 UTs, so it should not be a problem for now, but https://github.com/elastic/apm-pipeline-library/pull/1417 will be the one reducing the load quite a bit.

**Comment by trentm** [comment]
I mean it is a blocker, because I do not want to merge a change that makes build times on this repo go way up. Especially just to report junit-format tests that are not used by @astorm and I.

> As a workaround, If you'd like to speed up your development, you could comment out the notifyBuildResult step in the Jenkinsfile.

Does that disable updating the GitHub "check" status on PRs?

**Comment by v1v** [comment]
Gotcha!
Somehow I thought this PR was a draft PR and it was blocking your development. That's the reason I mentioned about the workaround for speeding up the development while using the CI.

You won't need to do much, since we just merged the PR with the fix, so I'll release a new version soon and you will be able to test it with the fix. Sorry for taking a bit of time

**Comment by trentm** [comment]
@v1v  Oh! I hadn't realized that https://github.com/elastic/apm-pipeline-library/pull/1417 is expected to fix slow `curl -o tests-info.json ...` issue.

**Comment by trentm** [comment]
^^^ that's much better:  https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-nodejs%2Fapm-agent-nodejs-mbp/detail/PR-2478/10/pipeline

The `#!/bin/bash -x chmod 755 generate-build-data.sh ./generate-build-data.sh https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/apm-agent-nodejs-mbp/PR-2478/ https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/apm-agent-nodejs-mbp/PR-2478/runs/10 SUCCESS 1269053 — generate-build-data` took `6s`.

Instead of the huge `tests-info.json` it is now downloading a `tests-summary.json` which is nice and small:

```
% curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/apm-agent-nodejs-mbp/PR-2478/runs/10/blueTestSummary/ -s | json
{
  "_class": "io.jenkins.blueocean.rest.model.BlueTestSummary",
  "_links": {
    "self": {
      "_class": "io.jenkins.blueocean.rest.hal.Link",
      "href": "/blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-2478/runs/10/blueTestSummary/"
    }
  },
  "existingFailed": 0,
  "failed": 0,
  "fixed": 0,
  "passed": 227704,
  "regressions": 0,
  "skipped": 0,
  "total": 227704
}
```

**Review by astorm — COMMENTED** [review]
Looks fine to me and no objections, but I'm going to leave approval to the robots folks, as this is code I don't really work in often and don't understand its implicit/explicit assumptions.

**Comment by trentm** [comment]
> Looks fine to me and no objections, but I'm going to leave approval

@astorm Thanks. Yah, I mostly wanted to make sure you were aware of the changes and didn't have any objections.

File: .ci/scripts/docker-test.sh [hunk 1]
```diff
@@ -54,8 +54,14 @@ node --version
 npm --version
 npm_install
 
+# Attempt to provide xunit-formatted test results, for Jenkins' "Test Results"
+# and other features like flaky-test reporting.
 if [[ -n ${TAV} ]]; then
-  npm run test:tav|tee tav-output.tap
+  npm run test:tav
+  # Currently the TAV tests do not support TAP or xunit-formatted output.
```
Comment (v1v) [inline .ci/scripts/docker-test.sh:61]:
```suggestion
  # Currently the TAV tests do not support TAP or junit-formatted output.
```

File: .ci/scripts/docker-test.sh [hunk 2]
```diff
@@ -54,8 +54,14 @@ node --version
 npm --version
 npm_install
 
+# Attempt to provide xunit-formatted test results, for Jenkins' "Test Results"
```
Comment (v1v) [inline .ci/scripts/docker-test.sh:57]:
```suggestion
# Attempt to provide junit-formatted test results, for Jenkins' "Test Results"
```

**Review by v1v — APPROVED** [review]
Thanks for this! 🙇 
The title also contains `xunit`

**Pull request merged**